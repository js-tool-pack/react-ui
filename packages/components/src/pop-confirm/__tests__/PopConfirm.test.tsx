import { PopConfirm } from '..';
import { fireEvent, render } from '@testing-library/react';

describe('PopConfirm', () => {
  // 模拟 ResizeObserver，ResizeObserver 不存在于 jsdom 中
  const MockObserverInstance: ResizeObserver = {
    observe: jest.fn(),
    unobserve: jest.fn(),
    disconnect: jest.fn(),
  };
  beforeEach(() => {
    global.ResizeObserver = jest
      .fn()
      .mockImplementation(() => MockObserverInstance);
  });

  test('attrs', () => {
    const onClick = jest.fn();
    const { container } = render(
      <PopConfirm
        visible
        attrs={{ className: 'foo', style: { background: '#fff' }, onClick }}
        appendTo={null}>
        <div>foo bar</div>
      </PopConfirm>,
    );

    const balloon = container.querySelector('.t-word-balloon');

    expect(balloon).toHaveClass('foo');
    expect(balloon).toHaveStyle({ background: '#fff' });

    expect(onClick).not.toBeCalled();
    fireEvent.click(balloon!);
    expect(onClick).toBeCalled();
  });

  test('icon', () => {
    const { container } = render(
      <PopConfirm visible icon="foo-bar" content="bar" appendTo={null}>
        <div>foo</div>
      </PopConfirm>,
    );
    const balloon = container.querySelector('.t-word-balloon');
    expect(balloon).toMatchSnapshot();
  });

  describe('callback', () => {
    const getBalloon = () => document.querySelector('.t-word-balloon');
    test('cancel undo', () => {
      const onCancel = jest.fn(() => false);
      render(
        <div>
          <PopConfirm
            visible
            destroyOnHide
            icon="foo-bar"
            onCancel={onCancel}
            content="bar">
            <div>foo</div>
          </PopConfirm>
          <div className="hide">hide</div>
        </div>,
      );
      // expect(getBalloon()).toMatchSnapshot();
      expect(onCancel).not.toBeCalled();
      fireEvent.click(document.querySelector('.t-pop-confirm__cancel')!);
      expect(getBalloon()).not.toHaveClass('t-pop-confirm-leave-active');
      expect(onCancel).toBeCalled();
    });
    test('cancel', () => {
      const onCancel = jest.fn();
      render(
        <div>
          <PopConfirm
            visible
            destroyOnHide
            icon="foo-bar"
            onCancel={onCancel}
            content="bar">
            <div>foo</div>
          </PopConfirm>
          <div className="hide">hide</div>
        </div>,
      );
      // expect(getBalloon()).toMatchSnapshot();
      expect(onCancel).not.toBeCalled();
      fireEvent.click(document.querySelector('.t-pop-confirm__cancel')!);
      expect(getBalloon()).toHaveClass('t-pop-confirm-leave-active');
      expect(onCancel).toBeCalled();
    });
    test('confirm undo', () => {
      const onConfirm = jest.fn(() => false);
      render(
        <div>
          <PopConfirm
            visible
            destroyOnHide
            icon="foo-bar"
            onConfirm={onConfirm}
            content="bar">
            <div>foo</div>
          </PopConfirm>
          <div className="hide">hide</div>
        </div>,
      );
      // expect(getBalloon()).toMatchSnapshot();
      expect(onConfirm).not.toBeCalled();
      fireEvent.click(document.querySelector('.t-pop-confirm__confirm')!);
      expect(getBalloon()).not.toHaveClass('t-pop-confirm-leave-active');
      expect(onConfirm).toBeCalled();
    });
    test('confirm', () => {
      const onConfirm = jest.fn();
      render(
        <div>
          <PopConfirm
            visible
            destroyOnHide
            icon="foo-bar"
            onConfirm={onConfirm}
            content="bar">
            <div>foo</div>
          </PopConfirm>
          <div className="hide">hide</div>
        </div>,
      );
      // expect(getBalloon()).toMatchSnapshot();
      expect(onConfirm).not.toBeCalled();
      fireEvent.click(document.querySelector('.t-pop-confirm__confirm')!);
      expect(getBalloon()).toHaveClass('t-pop-confirm-leave-active');
      expect(onConfirm).toBeCalled();
    });
  });
});
