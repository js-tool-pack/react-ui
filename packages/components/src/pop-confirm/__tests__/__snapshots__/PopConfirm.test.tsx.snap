// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopConfirm icon 1`] = `
<div
  class="t-word-balloon t-pop-confirm t-word-balloon--top t-pop-confirm-enter-active t-pop-confirm-enter-to"
  style="top: -10px; left: 0px;"
>
  <div
    class="t-word-balloon__content"
  >
    <section
      class="t-layout t-pop-confirm__layout t-layout--v"
    >
      <section
        class="t-main t-pop-confirm__main"
      >
        <div
          class="t-space t-pop-confirm__space"
          style="--t-space-fill-ratio: 100%;"
        >
          foo-bar
          <div
            class="t-pop-confirm__content"
          >
            bar
          </div>
        </div>
      </section>
      <section
        class="t-footer t-pop-confirm__footer"
      >
        <section
          class="t-space"
          style="--t-space-fill-ratio: 100%;"
        >
          <button
            class="t-button t-pop-confirm__cancel t-button--type-info t-button--icon-l t-button--plain t--size-sm"
          >
            <span>
              取消
            </span>
          </button>
          <button
            class="t-button t-pop-confirm__confirm t-button--type-primary t-button--icon-l t--size-sm"
          >
            <span>
              确认
            </span>
          </button>
        </section>
      </section>
    </section>
  </div>
  <div
    class="t-word-balloon__arrow"
  >
    <svg
      viewBox="-8 -5.5 16 5.515"
    >
      <path
        d="M8-5.5A4 4 180 005.1716-4.3284L1.4142-.5711A2 2 180 01-1.4142-.5711L-5.1716-4.3284A4 4 180 00-8-5.5Z"
      />
    </svg>
  </div>
</div>
`;
